src/main/java/com/celements/parents/DocumentParentsLister.java

```diff
@@ -21,61 +29,94 @@
-            ).getDocumentParentsList(docRef);
-        if (parentList.size() > 0) {
-          joinParentLists(docParents, parentList);
-          docRef = docParents.get(docParents.size() - 1);
-        }
-        hasMore = false;
-        if (secondKeys.size() > 0) {
-          int secKeyIndex = -1;
-          parentList = Collections.emptyList();
-          while (((secKeyIndex + 1) < secondKeys.size()) && ((parentList == null)
-              || parentList.isEmpty())) {
-            secKeyIndex = secKeyIndex + 1;
-            parentList = docParentProviderMap.get(secondKeys.get(secKeyIndex)
-                ).getDocumentParentsList(docRef);
-          }
-          hasMore = ((parentList != null) && !parentList.isEmpty());
-          if (hasMore) {
-            joinParentLists(docParents, parentList);
-            docRef = docParents.get(docParents.size() - 1);
-            secondKeys = new ArrayList<String>(secondKeyMaster);
-            secondKeys.remove(secKeyIndex);
-          }
-        }
-      } while (hasMore);
     } catch (XDocRecursionException recExp) {
       _LOGGER.info("Recursion in document parents found [" + recExp + "].");
     }
-    return docParents;
+    return parents;
+  }
+
+  public void setXDocParents(List<DocumentReference> ret) throws XDocRecursionException {
+    DocumentReference docRef = Iterables.getLast(ret);
+    for (DocumentReference parentRef : xDocParents.getDocumentParentsList(docRef)) {
+      setParent(ret, parentRef);
+    }
+  }
+
+  public boolean setNextProviderParent(List<DocumentReference> parents, 
+      Set<String> providers) throws XDocRecursionException {
+    DocumentReference docRef = Iterables.getLast(parents);
+    List<DocumentReference> nextParents = Collections.emptyList();
+    String provider = null;
+    Iterator<String> iter = providers.iterator();
+    while (nextParents.isEmpty() && iter.hasNext()) {
+      provider = iter.next();
+      nextParents = docParentProviderMap.get(provider).getDocumentParentsList(docRef);
+      nextParents = checkPageTypes(nextParents);
+    }
+    if (!nextParents.isEmpty()) {
+      providers.clear();
+      providers.addAll(getProvidersWithout(provider));
+      setParent(parents, nextParents.get(0));
+      if (nextParents.size() > 1) {
+        _LOGGER.warn("Received multiple parents for '{}' from provider '{}': {}", docRef,
+            provider, nextParents);
+      }
+      return true;
+    } else {
+      return false;
+    }
   }
 
-  private void joinParentLists(ArrayList<DocumentReference> docParents,
-      List<DocumentReference> parentList) throws XDocRecursionException {
-    for (DocumentReference parentRef : parentList) {
-      if (!docParents.contains(parentRef)) {
-        docParents.add(parentRef);
-      } else {
-        throw new XDocRecursionException(parentRef);
+  private Set<String> getProvidersWithout(String provider) {
+    Set<String> ret = new HashSet<>(docParentProviderMap.keySet());
```

Comment:
I don't think we have ever discussed a code style regarding automatic type inference since it is a Java 7 feature and we only recently switched to Java 7.